### ISSUE TITLE
TypeError: 'unicode' does not have the buffer interface

### ISSUE THREAD

**Issue by iaroslav-ai** [bokeh/bokeh#6653]
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

bokeh 0.12.6
python 2.7.12
64 bit Ubuntu 16.04
Chrome browser

#### Description of expected behavior and the observed behavior

I get `500: Internal Server Error` when I try to use signed sessions with bokeh. 

#### Complete, minimal, self-contained example code that reproduces the issue

I use `sliders.py` example from Bokeh Gallery. I use the following to run the app and generate the session id:

Run the app:
```bash
export BOKEH_SECRET_KEY=KKXYhtUJmQ5f8LiQJdeUGOF4HIeTzf5zOBZmlxrXtF6D
export BOKEH_SIGN_SESSIONS=True
bokeh serve sliders.py --session-ids external-signed
```

Generate session id:
```bash 
export BOKEH_SECRET_KEY=KKXYhtUJmQ5f8LiQJdeUGOF4HIeTzf5zOBZmlxrXtF6D
export BOKEH_SIGN_SESSIONS=True
python -c "from bokeh.util.session_id import generate_session_id; print(generate_session_id())"
```

Example url to access app:
http://localhost:5006/sliders?bokeh-session-id=4RJKVrnFVe60gB5urh9sE3jUnSGDkJAfCwvoaDsoMB8f-W6QAfyDoxORtN7mb6DHAzftAhpfnxVdzC-6gIT13uV0

#### Stack traceback and/or browser JavaScript console output

JavaScript console output:
```
Failed to load resource: the server responded with a status of 500 (Internal Server Error)
```

Bokeh traceback:
```
2017-07-21 00:23:58,161 Uncaught exception GET /sliders?bokeh-session-id=4RJKVrnFVe60gB5urh9sE3jUnSGDkJAfCwvoaDsoMB8f-W6QAfyDoxORtN7mb6DHAzftAhpfnxVdzC-6gIT13uV0 (::1)
HTTPServerRequest(protocol='http', host='localhost:5006', method='GET', uri='/sliders?bokeh-session-id=4RJKVrnFVe60gB5urh9sE3jUnSGDkJAfCwvoaDsoMB8f-W6QAfyDoxORtN7mb6DHAzftAhpfnxVdzC-6gIT13uV0', version='HTTP/1.1', remote_ip='::1', headers={'Accept-Language': 'en-US,en;q=0.8,ru;q=0.6,de;q=0.4', 'Accept-Encoding': 'gzip, deflate, br', 'Host': 'localhost:5006', 'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/59.0.3071.109 Chrome/59.0.3071.109 Safari/537.36', 'Connection': 'keep-alive', 'Cookie': '_xsrf=2|4ab25a11|f7d3fbdf1fab4d2b01135d63518a4f9a|1498856594; username-localhost-8888="2|1:0|10:1499014969|23:username-localhost-8888|44:ODVmNmU2NjIwYjUwNDlhYzk2MzY4OWQ5NDU2ZTExYjU=|3a908d5ba83bca558deae2665732f340eeef5ce69a2763c6cef367fd892e22b7"', 'Upgrade-Insecure-Requests': '1'})
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/web.py", line 1469, in _execute
    result = yield result
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1015, in run
    value = future.result()
  File "/usr/local/lib/python2.7/dist-packages/tornado/concurrent.py", line 237, in result
    raise_exc_info(self._exc_info)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1021, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/local/lib/python2.7/dist-packages/bokeh/server/views/doc_handler.py", line 27, in get
    session = yield self.get_session()
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 1015, in run
    value = future.result()
  File "/usr/local/lib/python2.7/dist-packages/tornado/concurrent.py", line 237, in result
    raise_exc_info(self._exc_info)
  File "/usr/local/lib/python2.7/dist-packages/tornado/gen.py", line 285, in wrapper
    yielded = next(result)
  File "/usr/local/lib/python2.7/dist-packages/bokeh/server/views/session_handler.py", line 36, in get_session
    signed=self.application.sign_sessions):
  File "/usr/local/lib/python2.7/dist-packages/bokeh/util/session_id.py", line 156, in check_session_id_signature
    return hmac.compare_digest(expected_signature, provided_signature)
TypeError: 'unicode' does not have the buffer interface
```

#### Screenshots or screencasts of the bug in action

![eg](https://user-images.githubusercontent.com/14038787/28441559-a70ff970-6dab-11e7-9ff0-c0218679bfbf.png)

**Comment by iaroslav-ai** [comment]
P.S. if I set `export BOKEH_SIGN_SESSIONS=False` I do not get any exceptions, but Bokeh says that session id is Invalid (error 403) no matter what I do.

**Comment by bryevdv** [comment]
@iaroslav-ai thanks for the detailed report. It's possible that simply calling `bokeh.util.string.encode_utf8` on both the arguments to `compare_digest` will solve the issue. Can you give that a try and report back?

**Comment by iaroslav-ai** [comment]
Hmm could not yet set the development version of bokeh. Will try again in few days.

**Comment by bryevdv** [comment]
I would actually suggest just editing the file in your site packages as a quicker way to test.

**Comment by iaroslav-ai** [comment]
@bryevdv that is a nice suggestion. Tried just that a moment ago and it seems to work now when I use  `encode_utf8` :D 
Should I close this or should it be open until 0.12.8?

**Comment by bryevdv** [comment]
If you have the ability to submit a quick PR with the change I can test it easily on python 3.4+ then we could get it merged for 0.12.7. The core team is pretty underwater with other tasks right now so submitting small PR's like this helps reduce the overhead burden quite a lot.

**Comment by iaroslav-ai** [comment]
Yeah lets do this while I am at it
